index.bs

```diff
@@ -0,0 +1,221 @@
+* <dfn>previousFrameStartingPoint</dfn> which is initially set to <code>null</code>.
+* <dfn>currentVisualRepresentation</dfn> which is initially set to <code>null</code>.
+* <dfn>previousVisualRepresentation</dfn> with is initially set to <code>null</code>.
+
+<div algorithm="additions to update rendering">
+    In the <a>update the rendering</a> step of the <a>event loop processing model</a>, add a step right after the step that calls <a>mark paint timing</a>:
+
+    1. Invoke the algorithm to <a>evaluate the layout instability</a> of a document with the complete set of <a>element</a>s for that {{Document}}.
+</div>
+
+
+Evaluate Layout Instability {#sec-eval-layout-instability}
+--------------------------------------------------------
+
+<div algorithm="evaluate the layout instability">
+    When asked to <dfn export>evaluate the layout instability</dfn>, with <var>elements</var> as the list of elements to be evaluated, run the following steps:
+
+    1. Let <var>unstable elements</var> be an empty list.
+    1. Let <var>total width</var> be the current <a>viewport</a> width.
+    1. Let <var>total height</var> be the current <a>viewport</a> height. 
+    1. For each <var>element</var> of <var>elements</var>:
+        1. Set <a>currentFrameStartingPoint</a> to the <a>starting point</a> of <var>element</var>.
+        1. Set <a>currentVisualRepresentation</a> to the <a>visual representation</a> of <var>element</var>.
+        1. Let <var>unstable</var> be the boolean returned from calling <a>identify an unstable element</a> with <var>element</var>’s <a>currentFrameStartingPoint</a> and <a>previousFrameStartingPoint</a>.  
+        1. If <var>unstable</var> is true, add a tuple to <var>unstable elements</var> with:
+            1. <a>currentVisualRepresentation</a> of <var>element</var>
+            1. <a>previousVisualRepresentation</a> of <var>element</var>
+            1. <var ignore=''>shiftFraction</var> of <var>element</var>:
+                1. Set to 0 if <a>previousFrameStartingPoint</a> is null.
+                1. Set to the max(abs(<a>previousFrameStartingPoint</a>.height - <a>currentFrameStartingPoint</a>.height) / <var>total height</var> , abs(<a>previousFrameStartingPoint</a>.width - <a>currentFrameStartingPoint</a>.width) / <var>total width</var>).
+        1. Set <a>previousFrameStartingPoint</a> point to <a>currentFrameStartingPoint</a>.
+        1. Set <a>previousVisualRepresentation</a> to the <a>currentVisualRepresentation</a>.
+    1. Call <a>compute the layout shift</a> with <var>unstable elements</var>.
+</div>
+
+Compute the Layout Shift {#sec-compute-layout-shift}
+-----------------------------------------------------
+
+<div algorithm="compute the layout shift">
+    When asked to to <dfn export>compute the layout shift</dfn>, with <var>unstable elements</var>, a list of impacted <a>element</a>'s on the page, as input, run the following steps:
+
+    1. Let <var>impacted region</var> be initially set to empty.
```

Comment:
Should this be "empty list" or some other data structure? I'm not sure if the term "empty" carries explicit meaning here.

Reply:
So I am thinking it would null.  This is a region type so until it is computed and set I think it would be null.

Reply:
@skobes 

Adding steve in case he has a different thought.

Reply:
"Empty" seems clearer to me, but I don't think it makes much difference.

Reply:
In that case, to avoid confusion with something like empty list, worth linking to a description of what an empty 'CSS box' means (or whichever is the type of rect being used here).

Reply:
The region represents a union of n rectangles.  So an empty region is not dissimilar from an empty list.

Reply:
I think the problem is that "empty" is ambiguous. Is it an empty list, empty set, empty string, empty rectangle? This is the algorithm. It should be clear whether impacted region is a list of rectangles or a single rectangle.

Reply:
Empty set of rectangles?  If you think plural rectangles is better than single empty region I think this makes sense.

Reply:
I think the spec should invoke (and explain, to the degree we think necessary) the mathematical concept of a region without mandating (or even necessarily suggesting) any particular implementation of that concept.

"Let impactedRegion be a two-dimensional geometric region which is initially empty (that is, it contains no points)."
[...]
"If previous visual representation is not null, expand impactedRegion to additionally contain all points that lie within tuple's current visual representation, tuple's previous visual representation, or both."

How does that sound?

Reply:
Since this is not exposed (only used to compute the fraction), then I think that would be ok. Alternatively we can use DOMRect as the type.

Reply:
Done.